feat: simplify strategy to current treasury decision, workaround for n-bpts and eip170

petrovska-petro opened the pull request:
brings down size to 18.9kB without optimizer enable while compiling. how?

- adapt the contract to current logic of the treasury decision. i.e: assume always selling of bal and full aura locking in behalf of voter. keeping the flexi for the aura breakdown

- include 3 *EnumerableSet* to support n-bpts and modify contract across board to adapt to the set lib

- contract is upgradable, always can be changed with an upgrade if treasury decides fits needs

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -136,20 +133,14 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-
-        baseRewardPool1 = IBaseRewardPool(crvRewards1);
-        baseRewardPool2 = IBaseRewardPool(crvRewards2);
+    /// @param _pids Pool ID of tokens involved in the avatar
+    constructor(uint256[] memory _pids) {
+        for (uint256 i = 0; i < _pids.length; i++) {
+            pids.add(_pids[i]);
+            (address lpToken,,, address crvRewards,,) = AURA_BOOSTER.poolInfo(_pids[i]);
+            assets.add(lpToken);
+            baseRewardPools.add(crvRewards);
+        }
```
Remove `constructor` and move to `initialize`. State variables can't be initialized in constructors for proxy contracts

petrovska-petro replied:
moved to the *initialize*  in 1bf3520289ab8590692162543359a3039d54602b

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -183,9 +171,11 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-        asset2.safeApprove(address(AURA_BOOSTER), type(uint256).max);
+        // Booster approval for bpts
+        uint256 length = assets.length();
+        for (uint256 i = 0; i < length; i++) {
```
```suggestion
        for (uint256 i; i < length; ++i) {
```

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -429,49 +393,46 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-        if (_amountAsset1 > 0) {
-            asset1.safeTransferFrom(msg.sender, address(this), _amountAsset1);
-            AURA_BOOSTER.deposit(pid1, _amountAsset1, true);
+        for (uint256 i = 0; i < _pids.length; i++) {
```
```suggestion
        for (uint256 i; i < _pids.length; ++i) {
```

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -429,49 +393,46 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
+    /// @param _pids Pids target to stake into
+    /// @param _amountAssets Amount of assets to be staked.
+    function deposit(uint256[] memory _pids, uint256[] memory _amountAssets, bool initialDeposit) external onlyOwner {
+        if (initialDeposit) {
             // Initialize at first deposit
+            require(lastClaimTimestamp == 0, "AuraAvatarMultiToken: Already initialised");
             lastClaimTimestamp = block.timestamp;
         }
```
I think we can probably remove `initialDeposit` at the cost that an extra harvest will take place right after the first deposit

petrovska-petro replied:
removed

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -429,49 +393,46 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-        if (bptDeposited1 == 0 && bptDeposited2 == 0) {
+    /// @param _pids Pids target to stake into
+    /// @param _amountAssets Amount of assets to be staked.
+    function deposit(uint256[] memory _pids, uint256[] memory _amountAssets, bool initialDeposit) external onlyOwner {
```
Should we check if `_pids` are a subset of the pids in storage?

petrovska-petro replied:
checked on deposit and included an error for its case `PidNotIncluded(uint256 pid)` in 1bf3520289ab8590692162543359a3039d54602b

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -487,22 +448,43 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
+        require(pids.contains(_removePid), "AuraAvatarMultiToken: PID doesnt exist!");
+
+        (address lpToken,,, address crvRewards,,) = AURA_BOOSTER.poolInfo(_removePid);
+        uint256 stakedBal = IBaseRewardPool(crvRewards).balanceOf(address(this));
```
```suggestion
        uint256 stakedAmount = IBaseRewardPool(crvRewards).balanceOf(address(this));
```

petrovska-petro replied:
renamed in 1bf3520289ab8590692162543359a3039d54602b

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -514,34 +496,48 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
+
+        for (uint256 i = 0; i < length;) {
+            assetAmounts[i] = IBaseRewardPool(baseRewardPools.at(i)).balanceOf(address(this));
+            unchecked {
+                ++i;
+            }
```
Nice optimization, but we should keep this conssitent over all the for loops in the contract

petrovska-petro replied:
added the opti across the board in 1bf3520289ab8590692162543359a3039d54602b

shuklaayush commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -574,94 +574,60 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-
-        uint256 totalUsdcEarned = usdcEarnedFromBal + usdcEarnedFromAura;
 
         address ownerCached = owner();
         if (totalUsdcEarned > 0) {
             USDC.safeTransfer(ownerCached, totalUsdcEarned);
```
Guess we don't need to cache `owner` anymore

petrovska-petro replied:
removed in 1bf3520289ab8590692162543359a3039d54602b

shuklaayush reviewed (changes requested):
Dropped some comments. I guess we can also remove `swapAuraForUsdc`

petrovska-petro commented:
> Dropped some comments. I guess we can also remove `swapAuraForUsdc`

decided to keep it for flexi, since treasury may decided diff split along the way. initially decided to remove it cause space concerns but now we're at 18.9kb which is fine

petrovska-petro commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -532,18 +532,24 @@ contract AuraAvatarMultiToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtil
+            IBaseRewardPool baseRewardPool = IBaseRewardPool(baseRewardPools.at(i));
+
+            uint256 balEarned = baseRewardPool.earned(address(this));
+            uint256 balEarnedAdjusted = (balEarned * AURA_BOOSTER.getRewardMultipliers(address(baseRewardPool)))
```
confirmed the addition of new `reward multiplier` within the aura migration infra, which can limit the amount aura minted per `reward contract`, defined on the boster,  instead of the minting aura based on the 100% of bal rewards earned ✅ 

e.g: https://etherscan.io/address/0x4EFc8DED860Bc472FA8d938dc3fD4946Bc1A0a18#code#F32#L724

sajanrajdev commented on src/aura/AuraAvatarMultiToken.sol:
```diff
@@ -572,26 +562,37 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
-        if (_amountAsset2 > 0) {
-            baseRewardPool2.withdrawAndUnwrap(_amountAsset2, false);
-            asset2.safeTransfer(msg.sender, _amountAsset2);
+            IBaseRewardPool(crvRewards).withdrawAndUnwrap(_amountAssets[i], false);
```
Why not claiming rewards along the way? As it stands, should we withdraw with any outstanding rewards, these will be stuck there unless we call `claimRewardsAndSendToOwner()` which will require manual Msig work to process. If we were to claim them, anything outstanding would be transferred to the Avatar and process during the following harvest (or atomically should harvest be called manually).

Maybe I am missing something..

petrovska-petro replied:
it was a decision to keep separated the actions of deposit, wd, rewards claiming etc. As it stands now, even in the case that owner wds funds from specific position, as long as the sets contains the info around the base reward pool contract in storage, when the CL keeper will trigger internally `claimAndRegisterRewards` method, which will loop thru all the rewards avail, will claim them and process them as usual

sajanrajdev replied:
I see, perfect. I just confirmed that the `removeBptPositionInfo()` function will check if there are any rewards pending and claim them before removing the info, that would have been my only concern. Good stuff!
```
        // NOTE: verify pending rewards and claim. Processing is done separately
        if (baseRewardPool.earned(address(this)) > 0) {
            baseRewardPool.getReward();
        }
```

sajanrajdev commented on test/aura/AuraAvatarMultiTokenTest.t.sol:
```diff
@@ -47,76 +61,66 @@ contract AuraAvatarTwoTokenTest is Test, AuraAvatarUtils {
     ////////////////////////////////////////////////////////////////////////////
 
     function test_constructor() public {
-        assertEq(avatar.pid1(), PID_80BADGER_20WBTC);
```
nit but given that there is no constructor anymore maybe we should group all of this within the init test

petrovska-petro replied:
resolved in ed2ffa5b2116d7788bd4c529a75692aa7615deba

sajanrajdev commented on test/aura/AuraAvatarMultiTokenTest.t.sol:
```diff
@@ -153,26 +155,49 @@ contract AuraAvatarTwoTokenTest is Test, AuraAvatarUtils {
 
-        assertEq(address(avatarProxy.asset1()), address(BPT_80BADGER_20WBTC));
-        assertEq(address(avatarProxy.asset2()), address(BPT_40WBTC_40DIGG_20GRAVIAURA));
+    function test_pendingRewards() public {
```
Nice one!

sajanrajdev commented on test/aura/AuraAvatarMultiTokenTest.t.sol:
```diff
@@ -701,22 +751,105 @@ contract AuraAvatarTwoTokenTest is Test, AuraAvatarUtils {
+        avatar.addBptPositionInfo(21);
+    }
+
+    function test_addBptPositionInfo() public {
```
I would use this chance to test that the other sets also increased in size accordingly and a step further would be to checck that the new asset and rewardPool stored match those from the Booster for this PID

petrovska-petro replied:
added suggestions in the test in 20af3b87a884a027b7edeefff641fb286e1c5a05

sajanrajdev reviewed (commented):
## Aura Avatar Review
- Hardcodes the chainlink keeper registry address
- Allows for multiple BPTs to be managed (to be added via their PID upon init and more can be added in the future as needed)
- Allows for deposits of multiple BPTs simultaneously as an option with proper checks (BPTs must be added to the managed list first). Depositing will emit an event for each asset deposited with the corresponding address, amount and ts - LG
- WithdrawAll will withdraw all managed BPTs from their staked positions. It can now be also called by the manager (handy for emergency situations as the one experienced a few weeks ago)
- Modified the withdraw(amount) function to allow for spcefic PIDs to be withdrawn at specific amounts (also callable by manager),. This will also emit an event with the relevant info. - LG
- Adds functions to add/remove PIDs dynamically from the managed list. Removal requires that staked balance is 0
- Adjusted the uint type around to reduce contract size
- Introduced view functions for the different components of the managed positions including staked balances - LG
- `pendingRewards()` returns the total claimable balance of AURA and BAL from all BPTs added to any balance of these assets sitting in the contract. This is important because removing a BPT position from the managed list will claim its rewards into the contract without processing them, ie. there could be unprocessed AURA or BAL in the Avatar at any given time and should be accounted. Additionally, incorporated the new rewards Multiplier from the new Booster which alters the total amount of AURA to be rewarded. - LG
- Upkeep will be performed whenever the freq threshold is met and there's BAL to be claimed/sitting on the Avatar - LG
- Withdraw won't claim pending rewards - commented in line
- Modified the `_processRewards()` function to always sell 100% of BAL for USDC as per latest decisions
- Updated Booster address and removed unnecessary BAL processing addresses from old strategy

## Test Changes Review
- Tests were modified properly to reflect the changes from above
- Some comments dropped in-line

**NOTES:**
- Maybe a small concern but Aura allows for "extra rewards" to be added to the rewards pools which means that additional tokens could be received in the future by the Avatar. As it stands, we have no way to sweep these tokens as the `claimRewardsAndSendToOwner` will only transfer BAL and AURA. This applies in general to any other token stuck within the strat that is not either of these. Should we need to sweep, we will need to upgrade.

If the contract size is too constraint as it is to add a sweep function, maybe we should consider adding an `address additionalReward` parameter to the `claimRewardsAndSendToOwner()` and allow for it to transfer any other ERC20 as needed. Ofc, a general sweep function would be preferred, what do you think? @petrovska-petro @dapp-whisperer 

Amazing work over all! Just a few notes here and in-line

petrovska-petro commented on test/aura/AuraAvatarMultiTokenTest.t.sol:
```diff
@@ -491,200 +456,285 @@ contract AuraAvatarTwoTokenTest is Test, AuraAvatarUtils {
-        vm.expectEmit(true, false, false, true);
-        emit Withdraw(address(BPT_40WBTC_40DIGG_20GRAVIAURA), 20e18, block.timestamp);
-        avatar.withdraw(10e18, 20e18);
+        vm.startPrank(owner);
```
included test in 4c6cb0ac15cdca236a19250c8916e5326d605074

sajanrajdev replied:
Amazing

petrovska-petro commented:
> ## Aura Avatar Review
> 
>     * Hardcodes the chainlink keeper registry address
> 
>     * Allows for multiple BPTs to be managed (to be added via their PID upon init and more can be added in the future as needed)
> 
>     * Allows for deposits of multiple BPTs simultaneously as an option with proper checks (BPTs must be added to the managed list first). Depositing will emit an event for each asset deposited with the corresponding address, amount and ts - LG
> 
>     * WithdrawAll will withdraw all managed BPTs from their staked positions. It can now be also called by the manager (handy for emergency situations as the one experienced a few weeks ago)
> 
>     * Modified the withdraw(amount) function to allow for spcefic PIDs to be withdrawn at specific amounts (also callable by manager),. This will also emit an event with the relevant info. - LG
> 
>     * Adds functions to add/remove PIDs dynamically from the managed list. Removal requires that staked balance is 0
> 
>     * Adjusted the uint type around to reduce contract size
> 
>     * Introduced view functions for the different components of the managed positions including staked balances - LG
> 
>     * `pendingRewards()` returns the total claimable balance of AURA and BAL from all BPTs added to any balance of these assets sitting in the contract. This is important because removing a BPT position from the managed list will claim its rewards into the contract without processing them, ie. there could be unprocessed AURA or BAL in the Avatar at any given time and should be accounted. Additionally, incorporated the new rewards Multiplier from the new Booster which alters the total amount of AURA to be rewarded. - LG
> 
>     * Upkeep will be performed whenever the freq threshold is met and there's BAL to be claimed/sitting on the Avatar - LG
> 
>     * Withdraw won't claim pending rewards - commented in line
> 
>     * Modified the `_processRewards()` function to always sell 100% of BAL for USDC as per latest decisions
> 
>     * Updated Booster address and removed unnecessary BAL processing addresses from old strategy
> 
> 
> ## Test Changes Review
> 
>     * Tests were modified properly to reflect the changes from above
> 
>     * Some comments dropped in-line
> 
> 
> ## NOTES
> 
> Maybe a small concern but Aura allows for "extra rewards" to be added to the rewards pools which means that additional tokens could be received in the future by the Avatar. As it stands, we have no way to sweep these tokens as the `claimRewardsAndSendToOwner` will only transfer BAL and AURA. This applies in general to any other token stuck within the strat that is not either of these. Should we need to sweep, we will need to upgrade.
> 
> If the contract size is too constraint as it is to add a sweep function, maybe we should consider adding an `address additionalReward` parameter to the `claimRewardsAndSendToOwner()` and allow for it to transfer any other ERC20 as needed. Ofc, a general sweep function would be preferred, what do you think? @petrovska-petro @dapp-whisperer
> 
> Amazing work over all! Just a few notes here and in-line

Luckily the contract was sitting still at 19.4, giving around 5kB free. So managed to include diretly the erc20 sweeping function which directly sends tokens to the `owner` in 3fd24f3e3ff6b947116b0c1da190209aecb6c067

sajanrajdev reviewed (approved):
Fixes LGTM, amazing work!

dapp-whisperer commented:
This looks good to me, I don't really have anything to add to the previous reviews. I don't have any objections to the design and like the addition of sweep function.

I did have some general comments towards the future development path:

I see avatars as kind-of an API for the treasury to interact with a given protocol, the ability to support any Aura PID makes it more general-purpose which makes sense.

- confirmation of each new PID to track makes sense from an accounting / logging standpoint and I don't mind the explicit opt-in step.

- agree with withdrawAll having a lower confirmation threshold than full treasury vault, the treasury threshold has proven to be slower than we'd like in an emergency situation.

for the contract size limitations, I think there are some natural places to increase composability:
* pricing assets
* swap paths 
(basically the functions in AuraAvatarUtils, minus the actual swapping and approval actions)

I would also say constant addresses could go in a registry of sorts, but having them inline is notably nice for cost.

> Modified the _processRewards() function to always sell 100% of BAL for USDC as per latest decisions
In general we want to parameterize how we process tokens as much as possible and have the owner able to change this without upgrade. 

one thing to consider is keeping track of historical approvals if we're doing upgrades rather than migrations when a dependency address changes.
